Allow uploading of landing page image

tdonohue opened the pull request:
Fixes #285 

Adds a new "Settings > Appearances" Admin menu, which allows you to upload a custom banner image for the landing page.

jcoyne commented on app/uploaders/banner_image_uploader.rb:
```diff
@@ -0,0 +1,8 @@
+
+  # Define valid extensions for site banner image
+  def extension_white_list
+    %w(jpg jpeg png gif tif tiff)
```
TIFF isn't a format that most web browsers can display natively.  Even if it were so, it's not a good choice due to the file size.

tdonohue replied:
👍 Good point, will clean that up

tdonohue replied:
@mjgiarlo RE: dimensions... That's an area of question for me.  The current default (in Sufia) is an image that's 5000x1000 ([this one](https://cloud.githubusercontent.com/assets/92044/18370978/88ecac20-75f6-11e6-8399-6536640ef695.jpg)).  But, the [Lerna sample site](https://ggeisler.github.io/lerna-dashboard/sdr-home) has an image that is smaller, 2206x446 ([this stanford one](https://ggeisler.github.io/lerna-dashboard/sdr/stanford-oval-blurred.png)).  Either size seems to be "valid".

Currently, in my edit UI, I'm stating ["2000x500 (or larger)"](https://github.com/projecthydra-labs/lerna/pull/402/files#diff-6726cc707ff0afef2e3c00a03afeaa89R28) but have no code validation in place.  So, I guess the question in my mind is.. do we need to auto-resize, or should we just check if dimensions are *at least* a particular size?

mjgiarlo replied:
I defer to @jcoyne and @ggeisler on this one.

ggeisler replied:
5000px wide seems excessive to me (maybe not at some point in the future but for the near future, at least). No ordinary user is going to have a browser near that wide. And all other things being equal, an image with larger dimensions will be larger in filesize.

Resizing to something like 2200px seems reasonable to me. It would cover the vast majority of cases and in the rare case where a user has a wider viewport, I think the current relevant CSS will just scale the image up (not good for images in general but this will be a very rare case, I think).

Rather than requiring a minimum size, we could just have some help text associated with the file upload that says "Upload an image at least 2200 pixels wide for best results." If the user uploads something smaller and it doesn't look good because it's upscaled, that's their problem, we warned them.

jcoyne commented on app/uploaders/banner_image_uploader.rb:
```diff
@@ -0,0 +1,8 @@
+class BannerImageUploader < CarrierWave::Uploader::Base
+  include CarrierWave::Compatibility::Paperclip
```
Why do we care about Paperclip compatibility?

cbeer replied:
Do we need `CarrierWave::Compatibility::Paperclip`? I'm not seeing any other use of `paperclip` off-hand.

tdonohue replied:
To be honest, I only brought this in because Sufia's [`AvatarUploader`](https://github.com/projecthydra/sufia/blob/master/app/uploaders/sufia/avatar_uploader.rb
) does the same. I wasn't sure of it's importance, and forgot to ask.
 
Is there any situation where there could be multiple `Site` objects (and therefore multiple background images)?  I only ask because by default `CarrierWave` just stores all uploads under `public/uploads/`... I wasn't sure if there's any situation were we need these stored in different locations per `Site` (in which case we may need a custom storage path).

mjgiarlo replied:
@tdonohue yeah, we can expect to have multiple Sites per application instance. The storage question is a good one, and relates to https://github.com/projecthydra-labs/lerna/issues/149. You could potentially punt on that for now, and then take on 149 as a follow-on issue.

tdonohue replied:
This is also noted above by @jcoyne (see discussion there). Good catch though.

cbeer commented on app/models/site.rb:
```diff
@@ -3,6 +3,10 @@ class Site < ActiveRecord::Base
   validates :application_name, presence: true, allow_nil: true
 
+  # Allow for uploading of site's banner image
+  mount_uploader :banner_image, BannerImageUploader, mount_on: :banner_image_name
```
Can we rename the `banner_image_name` column back to `banner_image` and just use Carrierwave's defaults?

cbeer commented on app/models/site.rb:
```diff
@@ -24,4 +28,11 @@ def instance
+
+    def update_banner_config
+      # If a site banner_image uploaded, then update Sufia's banner_image to use it
+      Sufia.config.banner_image = banner_image? ? banner_image.url : nil
```
I'm not sure what `Sufia.config` is. Does that get persisted somewhere, or do we need to do this every request?

jcoyne replied:
this is an in-memory configuration.

tdonohue replied:
`Sufia.config` is how I'm updating Sufia's default configuration for its banner_image: https://github.com/projecthydra/sufia/blob/818cb4ca53d81fdb37835a0cd7c753d66eacd679/lib/sufia/configuration.rb#L10-L13

If there's a better way to achieve this, maybe I'm not aware of it (suggestions welcome!). Essentially, the goal here is to dynamically update Sufia's `banner_image` setting to the newly uploaded Lerna banner image location.

mjgiarlo replied:
Yeah, are we opening ourselves to weird cross-tenant contamination here, with multiple tenants relying on Sufia's in-memory configuration to set the background image?

tdonohue replied:
Crud.  You all are correct, `Sufia.config` is completely in-memory and therefore incorrect. If I simply restart the rails server this updated `banner_image` is not persisted.  

I need to find a different way of persisting this change, or I need to override the default Sufia landing page to use `@site.banner_image.url` by default.  Hints are welcome, but I'll also do some digging here.

cbeer commented on app/views/appearances/edit.html.erb:
```diff
@@ -0,0 +1,45 @@
+              </div>
+          <% end %>
+          <div class="field form-group">
+              <%= f.label :banner_image, '<span class="fa fa-camera"></span> Change site banner image'.html_safe %><br>
```
Is this using/can this use `simple_form`?

tdonohue replied:
Again, I was following the structure established in the other "edit" forms (e.g. `content_blocks/edit.html.erb` and `labels/edit.html.erb`).  Should all Lerna forms be using `simple_form` instead?

tdonohue replied:
I take it all back, @cbeer. 👍 Thanks for the hint on `simple_form`.  I hadn't used that before.. it's rather cool.  Updating this to page to just use `simple_form` which vastly simplifies the code.

cbeer commented on app/views/appearances/edit.html.erb:
```diff
@@ -0,0 +1,45 @@
+          <% end %>
+          <div class="field form-group">
+              <%= f.label :banner_image, '<span class="fa fa-camera"></span> Change site banner image'.html_safe %><br>
+              <%= image_tag @site.banner_image.url(), class: "img-responsive" if @site.banner_image? %>
```
Style thing -- lose the `()` from the method call..

cbeer commented on app/views/appearances/edit.html.erb:
```diff
@@ -0,0 +1,45 @@
+    <div class="panel panel-default">
+      <%= form_for(@site, url: site_appearances_path) do |f| %>
+        <div class="panel-heading">
+          <h3 class="panel-title">Site Appearance</h3>
```
I don't think we've been consistent, but ought this use the i18n string `sufia.admin.sidebar.appearance`?

tdonohue replied:
The `<h1>` above it is already using the i18n string `sufia.admin.sidebar.appearance`, which is why I didn't reuse it for this `<h3 class="panel-title">`.  I agree though that more of Lerna should use i18n strings. But, in this situation I was following the same structure as other "edit" pages (e.g. [`content_blocks/edit.html.erb`](https://github.com/projecthydra-labs/lerna/blob/master/app/views/content_blocks/edit.html.erb) and [`labels/edit.html.erb`](https://github.com/projecthydra-labs/lerna/blob/master/app/views/labels/edit.html.erb))

tdonohue commented:
@mjgiarlo , @jcoyne and @cbeer : This PR is now ready for (re-)review.  I believe I've addressed all the previous concerns. I've also added RSpec tests, and squashed all my extraneous commits into a single commit.  

Oh, and obviously, the functionality is working, and I'm now persisting the changed banner (by overriding the default Sufia homepage in Lerna). Feedback welcome!

jcoyne commented:
@tdonohue remove `[WIP]` from the title please

jcoyne commented:
I'd prefer we don't put a 1.1 MB object into the git repository. Can that be avoided?

tdonohue commented:
@jcoyne : good points. I can replace that test stanford image with a smaller one.  The code isn't actually validating the image is of a particular size (just recommends a width of 2200px or larger). So, RSpec shouldn't care.

tdonohue commented:
@jcoyne : I've just pushed up an update to use the same [test image.jpg that is in use by Sufia](https://github.com/projecthydra/sufia/blob/master/spec/fixtures/image.jpg).  This image is just 111kb in size (instead of the 1.1MB test image). RSpec tests were updated.

cbeer commented on app/views/layouts/homepage.html.erb:
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE html>
```
What's the difference between this and the layout upstream in sufia? It'd be useful to call that out with a comment when we need to update it in the future.

cbeer commented on app/views/layouts/homepage.html.erb:
```diff
@@ -0,0 +1,50 @@
+<%= render '/masthead' %>
+
+<div class="image-masthead">
+  <div class="background-container" style="background-image: url('<%= Site.instance.banner_image? ? Site.instance.banner_image.url : Sufia.config.banner_image %>')"></div>
```
Ah, I see the difference now. Maybe this would be well-suited to be a helper in sufia that we can override locally? That'd at least spare us the pain of overriding this whole layout..

mjgiarlo replied:
:+1: to making a change in Sufia that gives us a smaller seam for overriding in Lerna. Thanks for taking this on, @tdonohue!

cbeer commented on spec/controllers/appearances_controller_spec.rb:
```diff
@@ -0,0 +1,94 @@
+
+      context "site with existing banner image" do
+        before do
+          f = fixture_file_upload('/images/image.jpg', 'image/jpg')
```
I'd pick a different fixture image for copyright and branding reasons.

tdonohue replied:
Are there any you'd recommend? I was just going with what was also used by Sufia (i copied `image.jpg` over from Sufia).  Guess I'd need a hint of what would be deemed acceptable here.

tdonohue replied:
Understood. Again, I was just going with an image that was distributed by Sufia already... so, this should likely also get replaced in Sufia at some point: https://github.com/projecthydra/sufia/blob/master/spec/fixtures/image.jpg.  I'll look around for something tomorrow.

tdonohue commented:
Updated this PR to use a public domain image of the Hydra of Lerna (from The New York Public Library) in the RSpec tests. Specifically, this one: http://digitalcollections.nypl.org/items/71ac7570-2efe-0133-ed51-58d385a7b928
